File: chatuikit/src/main/java/com/cometchat/chatuikit/shared/views/suggestionlist/CometChatSuggestionList.java
package com.cometchat.chatuikit.shared.views.suggestionlist;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;

import androidx.annotation.ColorInt;
import androidx.annotation.Dimension;
import androidx.annotation.NonNull;
import androidx.annotation.StyleRes;
import androidx.recyclerview.widget.RecyclerView;

import com.cometchat.chatuikit.CometChatTheme;
import com.cometchat.chatuikit.R;
import com.cometchat.chatuikit.databinding.CometchatSuggestionListBinding;
import com.cometchat.chatuikit.shared.resources.utils.Utils;
import com.cometchat.chatuikit.shared.resources.utils.itemclicklistener.OnItemClickListener;
import com.cometchat.chatuikit.shared.resources.utils.recycler_touch.ClickListener;
import com.cometchat.chatuikit.shared.resources.utils.recycler_touch.RecyclerTouchListener;
import com.cometchat.chatuikit.shimmer.CometChatShimmerAdapter;
import com.cometchat.chatuikit.shimmer.CometChatShimmerUtils;
import com.google.android.material.card.MaterialCardView;

import java.util.List;

/**
 * A custom view that represents a suggestion list using a
 * {@link MaterialCardView}. The {@code
 * CometChatSuggestionList} allows the display of suggestions with customizable
 * attributes such as background color, stroke color, corner radius, and item
 * styles. It also supports avatars and text styling for each suggestion item.
 * This class provides methods to configure the appearance of the list items,
 * handle item clicks, apply style attributes, and manage a shimmer effect
 * during loading.
 */
public class CometChatSuggestionList extends MaterialCardView {
    private static final String TAG = CometChatSuggestionList.class.getSimpleName();
    private CometchatSuggestionListBinding binding;

    private boolean showAvatar = true;

    private @ColorInt int suggestionListBackgroundColor;
    private @ColorInt int suggestionListStrokeColor;
    private @Dimension int suggestionListStrokeWidth;
    private @Dimension int suggestionListCornerRadius;
    private @StyleRes int suggestionListItemAvatarStyle;
    private @StyleRes int suggestionListItemTextAppearance;
    private @ColorInt int suggestionListItemTextColor;
    private @ColorInt int suggestionListItemInfoTextColor;
    private @StyleRes int suggestionListItemInfoTextAppearance;
    private @Dimension int maxHeightLimit;

    private SuggestionListAdapter suggestionListAdapter;
    private OnItemClickListener<SuggestionItem> onItemClickListener;

    /**
     * Constructor to initialize the CometChatSuggestionList with a context.
     *
     * @param context the context in which the view is running
     */
    public CometChatSuggestionList(Context context) {
        this(context, null);
    }

    /**
     * Constructor to initialize the CometChatSuggestionList with a context and
     * attribute set.
     *
     * @param context the context in which the view is running
     * @param attrs   the set of attributes associated with the view
     */
    public CometChatSuggestionList(Context context, AttributeSet attrs) {
        this(context, attrs, R.attr.cometchatSuggestionListStyle);
    }

    /**
     * Constructor to initialize the CometChatSuggestionList with a context,
     * attribute set, and a default style attribute.
     *
     * @param context      the context in which the view is running
     * @param attrs        the set of attributes associated with the view
     * @param defStyleAttr the default style to apply to this view
     */
    public CometChatSuggestionList(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        if (!isInEditMode()) {
            inflateAndInitializeView(attrs, defStyleAttr);
        }
    }

    /**
     * Inflates and initializes the view by binding the layout and applying styles.
     *
     * @param attrs        the set of attributes associated with the view
     * @param defStyleAttr the default style to apply to this view
     */
    private void inflateAndInitializeView(AttributeSet attrs, int defStyleAttr) {
        binding = CometchatSuggestionListBinding.inflate(LayoutInflater.from(getContext()), this, true);
        Utils.initMaterialCard(this);
        init();
        applyStyleAttributes(attrs, defStyleAttr);
    }

    /**
     * Initializes the suggestion list adapter and configures the recycler view
     * listeners.
     */
    private void init() {
        suggestionListAdapter = new SuggestionListAdapter(getContext(), null);
        binding.recyclerViewSuggestionList.setAdapter(suggestionListAdapter);
        configureRecyclerViewListeners();
    }

    /**
     * Applies the style attributes defined in the XML or by default.
     *
     * @param attrs        the set of attributes associated with the view
     * @param defStyleAttr the default style to apply to this view
     */
    private void applyStyleAttributes(AttributeSet attrs, int defStyleAttr) {
        TypedArray directAttributes = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.CometChatSuggestionList, defStyleAttr, 0);
        @StyleRes int styleResId = directAttributes.getResourceId(R.styleable.CometChatSuggestionList_cometchatSuggestionListStyle, 0);
        directAttributes = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.CometChatSuggestionList, defStyleAttr, styleResId);
        extractAttributesAndApplyDefaults(directAttributes);
    }

    /**
     * Configures the item touch listener and scroll listener for the recycler view.
     */
    private void configureRecyclerViewListeners() {
        binding.recyclerViewSuggestionList.addOnItemTouchListener(new RecyclerTouchListener(getContext(),
                                                                                            binding.recyclerViewSuggestionList,
                                                                                            new ClickListener() {
                                                                                                @Override
                                                                                                public void onClick(@NonNull View view,
                                                                                                                    int position) {
                                                                                                    SuggestionItem suggestionItem = (SuggestionItem) view.getTag(
                                                                                                        R.string.cometchat_tag_item);
                                                                                                    if (onItemClickListener != null)
                                                                                                        onItemClickListener.OnItemClick(suggestionItem,
                                                                                                                                        position);
                                                                                                }

                                                                                                @Override
                                                                                                public void onLongClick(View view, int position) {
                                                                                                    SuggestionItem suggestionItem = (SuggestionItem) view.getTag(
                                                                                                        R.string.cometchat_tag_item);
                                                                                                    if (onItemClickListener != null)
                                                                                                        onItemClickListener.OnItemLongClick(
                                                                                                            suggestionItem,
                                                                                                            position);
                                                                                                }
                                                                                            }));

        binding.recyclerViewSuggestionList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                recyclerView.canScrollVertically(1);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
            }
        });
    }

    /**
     * Extracts style attributes from the provided TypedArray and applies default
     * values where necessary.
     *
     * @param typedArray the TypedArray containing style attributes
     */
    private void extractAttributesAndApplyDefaults(TypedArray typedArray) {
        if (typedArray == null) return;
        try {
            suggestionListBackgroundColor = typedArray.getColor(R.styleable.CometChatSuggestionList_cometchatSuggestionListBackgroundColor,
                                                                CometChatTheme.getBackgroundColor1(getContext()));
            suggestionListStrokeColor = typedArray.getColor(R.styleable.CometChatSuggestionList_cometchatSuggestionListStrokeColor,
                                                            CometChatTheme.getStrokeColorLight(getContext()));
            suggestionListStrokeWidth = typedArray.getDimensionPixelSize(R.styleable.CometChatSuggestionList_cometchatSuggestionListStrokeWidth, 0);
            suggestionListCornerRadius = typedArray.getDimensionPixelSize(R.styleable.CometChatSuggestionList_cometchatSuggestionListCornerRadius, 0);
            suggestionListItemAvatarStyle = typedArray.getResourceId(R.styleable.CometChatSuggestionList_cometchatSuggestionListItemAvatarStyle, 0);
            suggestionListItemTextAppearance = typedArray.getResourceId(R.styleable.CometChatSuggestionList_cometchatSuggestionListItemTextAppearance,
                                                                        0);
            suggestionListItemTextColor = typedArray.getColor(R.styleable.CometChatSuggestionList_cometchatSuggestionListItemTextColor,
                                                              CometChatTheme.getTextColorPrimary(getContext()));
            suggestionListItemInfoTextColor = typedArray.getColor(R.styleable.CometChatSuggestionList_cometchatSuggestionListItemInfoTextColor,
                                                                    CometChatTheme.getTextColorSecondary(getContext()));
            suggestionListItemInfoTextAppearance = typedArray.getResourceId(R.styleable.CometChatSuggestionList_cometchatSuggestionListItemInfoTextAppearance,
                                                                            0);
            updateUI();
        } finally {
            typedArray.recycle();
        }
    }

    /**
     * Updates the UI of the suggestion list based on the extracted style
     * attributes.
     */
    private void updateUI() {
        setSuggestionListBackgroundColor(suggestionListBackgroundColor);
        setSuggestionListStrokeColor(suggestionListStrokeColor);
        setSuggestionListStrokeWidth(suggestionListStrokeWidth);
        setSuggestionListCornerRadius(suggestionListCornerRadius);
        setSuggestionListItemAvatarStyle(suggestionListItemAvatarStyle);
        setSuggestionListItemTextAppearance(suggestionListItemTextAppearance);
        setSuggestionListItemTextColor(suggestionListItemTextColor);
        setSuggestionListItemInfoTextColor(suggestionListItemInfoTextColor);
        setSuggestionListItemInfoTextAppearance(suggestionListItemInfoTextAppearance);
    }

    /**
     * Sets the style of the suggestion list using a specified style resource.
     *
     * @param style the style resource to apply to the suggestion list
     */
    public void setStyle(@StyleRes int style) {
        if (style != 0) {
            TypedArray typedArray = getContext().getTheme().obtainStyledAttributes(style, R.styleable.CometChatSuggestionList);
            extractAttributesAndApplyDefaults(typedArray);
        }
    }

    /**
     * Returns the background color of the suggestion list.
     *
     * @return the background color of the suggestion list
     */
    public @ColorInt int getSuggestionListBackgroundColor() {
        return suggestionListBackgroundColor;
    }

    /**
     * Sets the background color of the suggestion list.
     *
     * @param suggestionListBackgroundColor the background color to be set
     */
    public void setSuggestionListBackgroundColor(@ColorInt int suggestionListBackgroundColor) {
        this.suggestionListBackgroundColor = suggestionListBackgroundColor;
        setCardBackgroundColor(suggestionListBackgroundColor);
    }

    /**
     * Returns the stroke color of the suggestion list.
     *
     * @return the stroke color of the suggestion list
     */
    public @ColorInt int getSuggestionListStrokeColor() {
        return suggestionListStrokeColor;
    }

    /**
     * Sets the stroke color of the suggestion list.
     *
     * @param suggestionListStrokeColor the stroke color to be set
     */
    public void setSuggestionListStrokeColor(@ColorInt int suggestionListStrokeColor) {
        this.suggestionListStrokeColor = suggestionListStrokeColor;
        setStrokeColor(suggestionListStrokeColor);
    }

    /**
     * Returns the stroke width of the suggestion list.
     *
     * @return the stroke width of the suggestion list
     */
    public @Dimension int getSuggestionListStrokeWidth() {
        return suggestionListStrokeWidth;
    }

    /**
     * Sets the stroke width of the suggestion list.
     *
     * @param suggestionListStrokeWidth the stroke width to be set
     */
    public void setSuggestionListStrokeWidth(@Dimension int suggestionListStrokeWidth) {
        this.suggestionListStrokeWidth = suggestionListStrokeWidth;
        setStrokeWidth(suggestionListStrokeWidth);
    }

    /**
     * Returns the corner radius of the suggestion list.
     *
     * @return the corner radius of the suggestion list
     */
    public @Dimension int getSuggestionListCornerRadius() {
        return suggestionListCornerRadius;
    }

    /**
     * Sets the corner radius of the suggestion list.
     *
     * @param suggestionListCornerRadius the corner radius to be set
     */
    public void setSuggestionListCornerRadius(@Dimension int suggestionListCornerRadius) {
        this.suggestionListCornerRadius = suggestionListCornerRadius;
        setRadius(suggestionListCornerRadius);
    }

    /**
     * Returns the avatar style of the suggestion list items.
     *
     * @return the avatar style resource of the suggestion list items
     */
    public @StyleRes int getSuggestionListItemAvatarStyle() {
        return suggestionListItemAvatarStyle;
    }

    /**
     * Sets the avatar style of the suggestion list items.
     *
     * @param suggestionListItemAvatarStyle the avatar style resource to be set
     */
    public void setSuggestionListItemAvatarStyle(@StyleRes int suggestionListItemAvatarStyle) {
        this.suggestionListItemAvatarStyle = suggestionListItemAvatarStyle;
        suggestionListAdapter.setSuggestionListItemAvatarStyle(suggestionListItemAvatarStyle);
    }

    /**
     * Returns the text appearance of the suggestion list items.
     *
     * @return the text appearance resource of the suggestion list items
     */
    public @StyleRes int getSuggestionListItemTextAppearance() {
        return suggestionListItemTextAppearance;
    }

    /**
     * Sets the text appearance of the suggestion list items.
     *
     * @param suggestionListItemTextAppearance the text appearance resource to be set
     */
    public void setSuggestionListItemTextAppearance(@StyleRes int suggestionListItemTextAppearance) {
        this.suggestionListItemTextAppearance = suggestionListItemTextAppearance;
        suggestionListAdapter.setSuggestionListItemTextAppearance(suggestionListItemTextAppearance);
    }

    /**
     * Returns the text color of the suggestion list items.
     *
     * @return the text color of the suggestion list items
     */
    public @ColorInt int getSuggestionListItemTextColor() {
        return suggestionListItemTextColor;
    }

    /**
     * Sets the text color of the suggestion list items.
     *
     * @param suggestionListItemTextColor the text color to be set
     */
    public void setSuggestionListItemTextColor(@ColorInt int suggestionListItemTextColor) {
        this.suggestionListItemTextColor = suggestionListItemTextColor;
        suggestionListAdapter.setSuggestionListItemTextColor(suggestionListItemTextColor);
    }

    /**
     * Sets the color resource ID for the suggestion list item info text.
     * @param suggestionListItemInfoTextColor the color resource ID to apply to the suggestion list item info text
     */
    public void setSuggestionListItemInfoTextColor(@ColorInt int suggestionListItemInfoTextColor) {
        this.suggestionListItemInfoTextColor = suggestionListItemInfoTextColor;
        suggestionListAdapter.setSuggestionListItemInfoTextColor(suggestionListItemInfoTextColor);
    }

    /**
     * Sets the style resource ID for the suggestion list item info text appearance.
     *
     * @param suggestionListItemInfoTextAppearance the style resource ID to apply to the suggestion list item info text appearance
     */
    public void setSuggestionListItemInfoTextAppearance(@StyleRes int suggestionListItemInfoTextAppearance) {
        this.suggestionListItemInfoTextAppearance = suggestionListItemInfoTextAppearance;
        suggestionListAdapter.setSuggestionListItemInfoTextAppearance(suggestionListItemInfoTextAppearance);
    }

    /**
     * Returns whether avatars are shown in the suggestion list.
     *
     * @return true if avatars are shown, false otherwise
     */
    public boolean showAvatar() {
        return showAvatar;
    }

    /**
     * Sets whether avatars should be shown in the suggestion list.
     *
     * @param showAvatar true to show avatars, false to hide them
     */
    public void showAvatar(boolean showAvatar) {
        this.showAvatar = showAvatar;
        suggestionListAdapter.showAvatar(showAvatar);
    }

    /**
     * Sets the maximum height limit for the suggestion list.
     *
     * @param maxHeightLimit the maximum height limit in pixels
     */
    public void setMaxHeightLimit(@Dimension int maxHeightLimit) {
        if (maxHeightLimit > 0) {
            this.maxHeightLimit = maxHeightLimit;
        }
    }

    /**
     * Measures the dimensions of the suggestion list, enforcing the maximum height
     * limit if set.
     *
     * @param widthMeasureSpec  the width measurement specification
     * @param heightMeasureSpec the height measurement specification
     */
    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);
        if (maxHeightLimit != 0 && heightSize > maxHeightLimit) {
            int heightMode = MeasureSpec.getMode(heightMeasureSpec);
            heightMeasureSpec = MeasureSpec.makeMeasureSpec(maxHeightLimit, heightMode);
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    /**
     * Sets the view holder listener for the suggestion list items.
     *
     * @param viewHolderListener the view holder listener to be set
     */
    public void setListItemView(SuggestionListViewHolderListener viewHolderListener) {
        suggestionListAdapter.setListItemView(viewHolderListener);
    }

    /**
     * Updates the suggestion list with a new list of suggestion items.
     *
     * @param suggestionItemList the new list of suggestion items
     */
    public void setList(List<SuggestionItem> suggestionItemList) {
        if (!suggestionItemList.isEmpty()) {
            showShimmer(false);
        }
        suggestionListAdapter.updateList(suggestionItemList);
    }

    /**
     * Shows or hides the shimmer effect on the suggestion list.
     *
     * @param visibility true to show shimmer, false to hide it
     */
    public void showShimmer(boolean visibility) {
        if (visibility) {
            CometChatShimmerAdapter adapter = new CometChatShimmerAdapter(1, R.layout.shimmer_cometchat_suggestion_list_items);
            binding.shimmerRecyclerViewSuggestion.setAdapter(adapter);
            binding.shimmerEffectFrame.setShimmer(CometChatShimmerUtils.getCometChatShimmerConfig(getContext()));
            binding.shimmerEffectFrame.startShimmer();
        } else {
            binding.shimmerEffectFrame.stopShimmer();
        }
        binding.recyclerViewSuggestionList.setVisibility(visibility ? GONE : VISIBLE);
        binding.shimmerEffectFrame.setVisibility(visibility ? VISIBLE : GONE);
    }

    /**
     * Sets the item click listener for the suggestion list items.
     *
     * @param onItemClickListener the item click listener to be set
     */
    public void setItemClickListener(OnItemClickListener<SuggestionItem> onItemClickListener) {
        if (onItemClickListener != null) {
            this.onItemClickListener = onItemClickListener;
        }
    }

    /**
     * Returns the binding object for the suggestion list.
     *
     * @return the binding object for the suggestion list
     */
    public CometchatSuggestionListBinding getBinding() {
        return binding;
    }
}

File: chatuikit-kotlin/src/main/java/com/cometchat/uikit/kotlin/presentation/shared/suggestionlist/CometChatSuggestionList.kt
package com.cometchat.uikit.kotlin.presentation.shared.suggestionlist

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import androidx.annotation.ColorInt
import androidx.annotation.Dimension
import androidx.annotation.StyleRes
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.cometchat.uikit.kotlin.R
import com.cometchat.uikit.kotlin.databinding.CometchatSuggestionListBinding
import com.cometchat.uikit.kotlin.presentation.shared.shimmer.CometChatShimmerAdapter
import com.cometchat.uikit.kotlin.presentation.shared.shimmer.CometChatShimmerUtils
import com.cometchat.uikit.kotlin.shared.formatters.SuggestionItem
import com.cometchat.uikit.kotlin.shared.resources.utils.Utils
import com.cometchat.uikit.kotlin.shared.resources.utils.itemclicklistener.OnItemClickListener
import com.cometchat.uikit.kotlin.shared.resources.utils.recycler_touch.ClickListener
import com.cometchat.uikit.kotlin.shared.resources.utils.recycler_touch.RecyclerTouchListener
import com.cometchat.uikit.kotlin.theme.CometChatTheme
import com.google.android.material.card.MaterialCardView

/**
 * CometChatSuggestionList is a custom view that displays a list of suggestions
 * for mentions or other autocomplete features.
 * 
 * Features:
 * - Shimmer loading state while fetching suggestions
 * - Item click handling for selection
 * - Max height limit to prevent taking too much screen space
 * - Custom item view support
 * - Scroll to bottom detection for pagination
 *
 * ## XML Usage
 *
 * ```xml
 * <com.cometchat.uikit.kotlin.presentation.shared.suggestionlist.CometChatSuggestionList
 *     android:id="@+id/suggestionList"
 *     android:layout_width="match_parent"
 *     android:layout_height="wrap_content"
 *     app:cometchatSuggestionListStyle="@style/CometChatSuggestionListStyle" />
 * ```
 *
 * ## Programmatic Usage
 *
 * ```kotlin
 * val suggestionList = CometChatSuggestionList(context)
 *
 * // Configure appearance
 * suggestionList.setMaxHeightLimit(250.dp)
 * suggestionList.showAvatar(true)
 * suggestionList.setSuggestionListBackgroundColor(Color.WHITE)
 * suggestionList.setSuggestionListCornerRadius(8.dp)
 *
 * // Set listeners
 * suggestionList.setItemClickListener(object : OnItemClickListener<SuggestionItem> {
 *     override fun OnItemClick(item: SuggestionItem, position: Int) {
 *         // Handle selection
 *     }
 *
 *     override fun OnItemLongClick(item: SuggestionItem, position: Int) {
 *         // Handle long press
 *     }
 * })
 *
 * suggestionList.setOnScrollToBottomListener {
 *     // Load more suggestions (pagination)
 * }
 *
 * // Update suggestions
 * suggestionList.setList(suggestions)
 *
 * // Show/hide loading state
 * suggestionList.showShimmer(true)
 * ```
 *
 * ## Style Configuration
 *
 * ```kotlin
 * val style = CometChatSuggestionListStyle(
 *     backgroundColor = Color.WHITE,
 *     strokeColor = Color.LTGRAY,
 *     strokeWidth = 1.dp,
 *     cornerRadius = 8.dp,
 *     maxHeight = 250.dp,
 *     itemTextColor = Color.BLACK,
 *     itemInfoTextColor = Color.GRAY
 * )
 * suggestionList.setStyle(style)
 * ```
 *
 * @see CometChatSuggestionListStyle Style configuration class
 * @see SuggestionItem Data model for suggestion items
 * @see SuggestionListAdapter Adapter for rendering items
 */
class CometChatSuggestionList @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = R.attr.cometchatSuggestionListStyle
) : MaterialCardView(context, attrs, defStyleAttr) {

    companion object {
        private val TAG = CometChatSuggestionList::class.java.simpleName
    }

    // View Binding
    private val binding: CometchatSuggestionListBinding

    // Adapter
    private val suggestionListAdapter: SuggestionListAdapter

    // Callbacks
    private var onItemClickListener: OnItemClickListener<SuggestionItem>? = null
    private var onScrollToBottomListener: (() -> Unit)? = null

    // Configuration
    private var showAvatar: Boolean = true
    @Dimension private var maxHeightLimit: Int = 0

    // Style properties
    @ColorInt private var suggestionListBackgroundColor: Int = 0
    @ColorInt private var suggestionListStrokeColor: Int = 0
    @Dimension private var suggestionListStrokeWidth: Int = 0
    @Dimension private var suggestionListCornerRadius: Int = 0
    @StyleRes private var suggestionListItemAvatarStyle: Int = 0
    @StyleRes private var suggestionListItemTextAppearance: Int = 0
    @ColorInt private var suggestionListItemTextColor: Int = 0
    @StyleRes private var suggestionListItemInfoTextAppearance: Int = 0
    @ColorInt private var suggestionListItemInfoTextColor: Int = 0

    init {
        binding = CometchatSuggestionListBinding.inflate(
            LayoutInflater.from(context), this, true
        )
        Utils.initMaterialCard(this)
        
        suggestionListAdapter = SuggestionListAdapter(context)
        binding.recyclerViewSuggestionList.adapter = suggestionListAdapter
        
        // Disable item animator to prevent flicker on list updates
        binding.recyclerViewSuggestionList.itemAnimator = null
        
        applyStyleAttributes(attrs, defStyleAttr)
        setupRecyclerViewListeners()
    }

    /**
     * Applies style attributes from XML.
     */
    private fun applyStyleAttributes(attrs: AttributeSet?, defStyleAttr: Int) {
        val directAttributes = context.theme.obtainStyledAttributes(
            attrs, R.styleable.CometChatSuggestionList, defStyleAttr, 0
        )
        
        @StyleRes val styleResId = directAttributes.getResourceId(
            R.styleable.CometChatSuggestionList_cometchatSuggestionListStyle, 0
        )
        
        val typedArray = context.theme.obtainStyledAttributes(
            attrs, R.styleable.CometChatSuggestionList, defStyleAttr, styleResId
        )
        
        extractAttributesAndApplyDefaults(typedArray)
    }

    /**
     * Extracts style attributes and applies defaults.
     */
    private fun extractAttributesAndApplyDefaults(typedArray: TypedArray) {
        try {
            suggestionListBackgroundColor = typedArray.getColor(
                R.styleable.CometChatSuggestionList_cometchatSuggestionListBackgroundColor,
                CometChatTheme.getBackgroundColor1(context)
            )
            suggestionListStrokeColor = typedArray.getColor(
                R.styleable.CometChatSuggestionList_cometchatSuggestionListStrokeColor,
                CometChatTheme.getBorderColorLight(context)
            )
            suggestionListStrokeWidth = typedArray.getDimensionPixelSize(
                R.styleable.CometChatSuggestionList_cometchatSuggestionListStrokeWidth,
                resources.getDimensionPixelSize(R.dimen.cometchat_1dp)
            )
            suggestionListCornerRadius = typedArray.getDimensionPixelSize(
                R.styleable.CometChatSuggestionList_cometchatSuggestionListCornerRadius,
                resources.getDimensionPixelSize(R.dimen.cometchat_corner_radius_2)
            )
            suggestionListItemAvatarStyle = typedArray.getResourceId(
                R.styleable.CometChatSuggestionList_cometchatSuggestionListItemAvatarStyle,
                R.style.CometChatAvatarStyle
            )
            suggestionListItemTextAppearance = typedArray.getResourceId(
                R.styleable.CometChatSuggestionList_cometchatSuggestionListItemTextAppearance,
                R.style.CometChatTextAppearanceHeading4_Medium
            )
            suggestionListItemTextColor = typedArray.getColor(
                R.styleable.CometChatSuggestionList_cometchatSuggestionListItemTextColor,
                CometChatTheme.getTextColorPrimary(context)
            )
            suggestionListItemInfoTextAppearance = typedArray.getResourceId(
                R.styleable.CometChatSuggestionList_cometchatSuggestionListItemInfoTextAppearance,
                R.style.CometChatTextAppearanceBody_Regular
            )
            suggestionListItemInfoTextColor = typedArray.getColor(
                R.styleable.CometChatSuggestionList_cometchatSuggestionListItemInfoTextColor,
                CometChatTheme.getTextColorSecondary(context)
            )
            
            updateUI()
        } finally {
            typedArray.recycle()
        }
    }

    /**
     * Updates the UI based on style properties.
     */
    private fun updateUI() {
        setCardBackgroundColor(suggestionListBackgroundColor)
        setStrokeColor(suggestionListStrokeColor)
        strokeWidth = suggestionListStrokeWidth
        radius = suggestionListCornerRadius.toFloat()
        
        suggestionListAdapter.setItemAvatarStyle(suggestionListItemAvatarStyle)
        suggestionListAdapter.setItemTextAppearance(suggestionListItemTextAppearance)
        suggestionListAdapter.setItemTextColor(suggestionListItemTextColor)
        suggestionListAdapter.setItemInfoTextAppearance(suggestionListItemInfoTextAppearance)
        suggestionListAdapter.setItemInfoTextColor(suggestionListItemInfoTextColor)
    }

    /**
     * Sets up RecyclerView listeners for item clicks and scroll detection.
     */
    private fun setupRecyclerViewListeners() {
        // Item click listener
        binding.recyclerViewSuggestionList.addOnItemTouchListener(
            RecyclerTouchListener(context, binding.recyclerViewSuggestionList, object : ClickListener {
                override fun onClick(view: View, position: Int) {
                    val item = view.getTag(R.string.cometchat_tag_item) as? SuggestionItem
                    item?.let { onItemClickListener?.OnItemClick(it, position) }
                }

                override fun onLongClick(view: View, position: Int) {
                    val item = view.getTag(R.string.cometchat_tag_item) as? SuggestionItem
                    item?.let { onItemClickListener?.OnItemLongClick(it, position) }
                }
            })
        )

        // Scroll listener for pagination
        binding.recyclerViewSuggestionList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (!recyclerView.canScrollVertically(1)) {
                    onScrollToBottomListener?.invoke()
                }
            }
        })
    }

    /**
     * Measures the view, enforcing max height limit if set.
     */
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var newHeightSpec = heightMeasureSpec
        if (maxHeightLimit > 0) {
            val heightSize = MeasureSpec.getSize(heightMeasureSpec)
            if (heightSize > maxHeightLimit) {
                val heightMode = MeasureSpec.getMode(heightMeasureSpec)
                newHeightSpec = MeasureSpec.makeMeasureSpec(maxHeightLimit, heightMode)
            }
        }
        super.onMeasure(widthMeasureSpec, newHeightSpec)
    }

    // ==================== Public API ====================

    /**
     * Sets the style using a style resource.
     */
    fun setStyle(@StyleRes style: Int) {
        if (style != 0) {
            val typedArray = context.theme.obtainStyledAttributes(
                style, R.styleable.CometChatSuggestionList
            )
            extractAttributesAndApplyDefaults(typedArray)
        }
    }

    /**
     * Sets the style using a CometChatSuggestionListStyle object.
     */
    fun setStyle(style: CometChatSuggestionListStyle) {
        suggestionListBackgroundColor = style.backgroundColor
        suggestionListStrokeColor = style.strokeColor
        suggestionListStrokeWidth = style.strokeWidth
        suggestionListCornerRadius = style.cornerRadius
        maxHeightLimit = style.maxHeight
        suggestionListItemAvatarStyle = style.itemAvatarStyle
        suggestionListItemTextAppearance = style.itemTextAppearance
        suggestionListItemTextColor = style.itemTextColor
        suggestionListItemInfoTextAppearance = style.itemInfoTextAppearance
        suggestionListItemInfoTextColor = style.itemInfoTextColor
        updateUI()
    }

    /**
     * Updates the suggestion list with new items.
     */
    fun setList(items: List<SuggestionItem>) {
        android.util.Log.d("MentionDebug", "[$TAG] setList() - received ${items.size} items: ${items.map { it.name }}")
        if (items.isNotEmpty()) {
            android.util.Log.d("MentionDebug", "[$TAG] setList() - hiding shimmer (items not empty)")
            showShimmer(false)
        }
        android.util.Log.d("MentionDebug", "[$TAG] setList() - calling adapter.updateList()")
        suggestionListAdapter.updateList(items)
    }

    /**
     * Shows or hides the shimmer loading state.
     */
    fun showShimmer(show: Boolean) {
        if (show) {
            // Use 1 shimmer item like Java implementation
            val shimmerAdapter = CometChatShimmerAdapter(
                1, R.layout.shimmer_cometchat_suggestion_list_items
            )
            binding.shimmerRecyclerView.adapter = shimmerAdapter
            binding.shimmerEffectFrame.setShimmer(
                CometChatShimmerUtils.getCometChatShimmerConfig(context)
            )
            binding.shimmerEffectFrame.startShimmer()
        } else {
            binding.shimmerEffectFrame.stopShimmer()
        }
        binding.recyclerViewSuggestionList.isVisible = !show
        binding.shimmerEffectFrame.isVisible = show
    }

    /**
     * Sets whether to show avatars in suggestion items.
     */
    fun showAvatar(show: Boolean) {
        showAvatar = show
        suggestionListAdapter.showAvatar(show)
    }

    /**
     * Sets the maximum height limit for the suggestion list.
     */
    fun setMaxHeightLimit(@Dimension maxHeight: Int) {
        if (maxHeight > 0) {
            maxHeightLimit = maxHeight
            requestLayout()
        }
    }

    /**
     * Sets the item click listener.
     */
    fun setItemClickListener(listener: OnItemClickListener<SuggestionItem>?) {
        onItemClickListener = listener
    }

    /**
     * Sets the scroll to bottom listener for pagination.
     */
    fun setOnScrollToBottomListener(listener: (() -> Unit)?) {
        onScrollToBottomListener = listener
    }

    /**
     * Sets a custom view holder listener for custom item views.
     */
    fun setListItemView(listener: SuggestionListViewHolderListener?) {
        suggestionListAdapter.setViewHolderListener(listener)
    }

    // ==================== Style Setters ====================

    fun setSuggestionListBackgroundColor(@ColorInt color: Int) {
        suggestionListBackgroundColor = color
        setCardBackgroundColor(color)
    }

    fun setSuggestionListStrokeColor(@ColorInt color: Int) {
        suggestionListStrokeColor = color
        setStrokeColor(color)
    }

    fun setSuggestionListStrokeWidth(@Dimension width: Int) {
        suggestionListStrokeWidth = width
        strokeWidth = width
    }

    fun setSuggestionListCornerRadius(@Dimension radius: Int) {
        suggestionListCornerRadius = radius
        this.radius = radius.toFloat()
    }

    fun setSuggestionListItemAvatarStyle(@StyleRes style: Int) {
        suggestionListItemAvatarStyle = style
        suggestionListAdapter.setItemAvatarStyle(style)
    }

    fun setSuggestionListItemTextAppearance(@StyleRes style: Int) {
        suggestionListItemTextAppearance = style
        suggestionListAdapter.setItemTextAppearance(style)
    }

    fun setSuggestionListItemTextColor(@ColorInt color: Int) {
        suggestionListItemTextColor = color
        suggestionListAdapter.setItemTextColor(color)
    }

    fun setSuggestionListItemInfoTextAppearance(@StyleRes style: Int) {
        suggestionListItemInfoTextAppearance = style
        suggestionListAdapter.setItemInfoTextAppearance(style)
    }

    fun setSuggestionListItemInfoTextColor(@ColorInt color: Int) {
        suggestionListItemInfoTextColor = color
        suggestionListAdapter.setItemInfoTextColor(color)
    }

    // ==================== Getters ====================

    fun getSuggestionListBackgroundColor(): Int = suggestionListBackgroundColor
    fun getSuggestionListStrokeColor(): Int = suggestionListStrokeColor
    fun getSuggestionListStrokeWidth(): Int = suggestionListStrokeWidth
    fun getSuggestionListCornerRadius(): Int = suggestionListCornerRadius
    fun getSuggestionListItemAvatarStyle(): Int = suggestionListItemAvatarStyle
    fun getSuggestionListItemTextAppearance(): Int = suggestionListItemTextAppearance
    fun getSuggestionListItemTextColor(): Int = suggestionListItemTextColor
    fun getSuggestionListItemInfoTextAppearance(): Int = suggestionListItemInfoTextAppearance
    fun getSuggestionListItemInfoTextColor(): Int = suggestionListItemInfoTextColor

    /**
     * Returns the binding for advanced customization.
     */
    fun getBinding(): CometchatSuggestionListBinding = binding
}
